api/src/clojure/org/akvo/flow_api/unilog/unilog.clj

```diff
@@ -39,21 +39,36 @@
 
 (def parse-json #(json/parse-string % true))
 
-(defn form-data [{:keys [formId id surveyId]}]
-  {:form-id formId
+(defn form-data [{:keys [formId id]}]
+  {:owner-id formId
    :id id})
 
 (defn form-instance-data [{:keys [formId formInstanceId]}]
-  {:form-id formId
+  {:owner-id formId
    :id formInstanceId})
 
 (defn datapoint-data [{:keys [id surveyId]}]
   {:id id
-   :survey-id surveyId})
+   :owner-id surveyId})
+
+(defn collapse [final {:keys [parent-delete-key parent-update-key child-delete-key child-update-key]}]
+  (let [parent-deleted (set (keep parent-delete-key final))
+        parent-updated (apply disj (set (keep parent-update-key final)) parent-deleted)
+        child-deleted (set (keep child-delete-key final))
```

Comment:
I understand that a `child-deleted` is required to derive the `child-changed`, but, is not the case that we shouldn't show a child-deleted if the parent has been deleted? So we need another _pass_ after calculating `child-changed`? :thinking:

Reply:
Client already has to deal with unknown deletes due to permissions.